### PR TITLE
[PART-III] TF-2802 Calendar event accept interactor and controller

### PR DESCRIPTION
**Pull request by tddang-linagora**
### Issue
- #2802

### CODE REVIEW

File: lib/features/email/domain/usecases/calendar_event_accept_interactor.dart [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import 'package:core/core.dart';
```
Comment (dab246) [inline lib/features/email/domain/usecases/calendar_event_accept_interactor.dart:1]: Only import class necessary

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 2]
```diff
@@ -1465,7 +1477,8 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
 
   void _handleParseCalendarEventSuccess(ParseCalendarEventSuccess success) {
     emailLoadedViewState.value = Right<Failure, Success>(success);
-    calendarEvent.value = success.calendarEventList.first;
+    calendarEvent.value = success.calendarEventMap.values.first.first;
+    _displayingEventBlobId = success.calendarEventMap.keys.first;
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:1481]: Please make sure `calendarEventMap` is not empty.

File: lib/features/email/domain/usecases/calendar_event_accept_interactor.dart (hunk 1 above)
Comment (tddang-linagora) [reply lib/features/email/domain/usecases/calendar_event_accept_interactor.dart]: Done

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 3]
```diff
@@ -1671,4 +1684,39 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
...
+        _acceptCalendarEventAction();
+        break;
+      default:
+        break;
+    }
+  }
+  
+  void _acceptCalendarEventAction() {
+    if (_acceptCalendarEventInteractor == null) return;
+    if (_displayingEventBlobId == null) return;
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:1700]:
```suggestion
    if (_acceptCalendarEventInteractor == null || _displayingEventBlobId == null) return;
```

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 4]
```diff
@@ -1671,4 +1684,39 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
...
+
+    consumeState(_acceptCalendarEventInteractor!.execute(
+      mailboxDashBoardController.accountId.value!,
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:1703]: Please also make sure `accountId ` is not null

File: lib/features/email/presentation/controller/single_email_controller.dart (hunk 2 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart:1481]: It is checked from `lib/features/email/domain/usecases/parse_calendar_event_interactor.dart` line `43 ~ 44`

File: lib/main/localizations/app_localizations.dart [hunk 5]
```diff
@@ -3844,4 +3844,18 @@ class AppLocalizations {
...
+    return Intl.message(
+      'Reply sent successfully!',
+      name: 'replySentSuccessfully',
+    );
+  }
+
+  String get replySentUnsuccessfully {
+    return Intl.message(
+      'Reply sent unsuccessfully!',
+      name: 'replySentUnsuccessfully',
```
Comment (dab246) [inline lib/main/localizations/app_localizations.dart:3858]: Please contact @hoangdat  to get suitable text

File: test/features/email/presentation/controller/single_email_controller_test.dart [hunk 6]
```diff
@@ -252,4 +261,39 @@ void main() {
...
   });
+
+  group('calendar event reply test:', () {
+    final blobId = Id('abc123');
+    final calendarEvent = CalendarEvent();
+
+    setUp(() {
+      when(mailboxDashboardController.selectedEmail).thenReturn(Rxn(null));
+      when(mailboxDashboardController.emailUIAction).thenReturn(Rxn(null));
+      when(mailboxDashboardController.viewState).thenReturn(Rx(Right(UIState.idle)));
```
Comment (dab246) [inline test/features/email/presentation/controller/single_email_controller_test.dart:272]: IMO, do not use `when` in `setUp` method

File: lib/features/email/presentation/controller/single_email_controller.dart (hunk 3 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

File: lib/features/email/presentation/controller/single_email_controller.dart (hunk 4 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

File: lib/main/localizations/app_localizations.dart (hunk 5 above)
Comment (tddang-linagora) [reply lib/main/localizations/app_localizations.dart]: Text replaced

File: test/features/email/presentation/controller/single_email_controller_test.dart (hunk 6 above)
Comment (tddang-linagora) [reply test/features/email/presentation/controller/single_email_controller_test.dart:272]: the test won't pass if I put it inside `test` function

File: lib/features/email/domain/state/parse_calendar_event_state.dart [hunk 7]
```diff
@@ -1,19 +1,20 @@
...
+import 'package:jmap_dart_client/jmap/core/id.dart';
 import 'package:jmap_dart_client/jmap/mail/calendar/calendar_event.dart';
 import 'package:tmail_ui_user/features/email/domain/model/event_action.dart';
 
 class ParseCalendarEventLoading extends LoadingState {}
 
 class ParseCalendarEventSuccess extends UIState {
 
-  final List<CalendarEvent> calendarEventList;
+  final Map<Id, List<CalendarEvent>> calendarEventMap;
```
Comment (hoangdat) [inline lib/features/email/domain/state/parse_calendar_event_state.dart:11]: should be 
```suggestion
  final Map<Id, CalendarEvent> calendarEventMap;
```

File: test/features/email/presentation/controller/single_email_controller_test.dart (hunk 6 above)
Comment (hoangdat) [reply test/features/email/presentation/controller/single_email_controller_test.dart:273]: should find the root cause. IMO, should use in `test` only

Comment (tddang-linagora) [reply test/features/email/presentation/controller/single_email_controller_test.dart:273]: Will try and fixup with update

**Review by hoangdat — APPROVED** [review]

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 8]
```diff
@@ -1465,7 +1478,8 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
 
   void _handleParseCalendarEventSuccess(ParseCalendarEventSuccess success) {
     emailLoadedViewState.value = Right<Failure, Success>(success);
-    calendarEvent.value = success.calendarEventList.first;
+    calendarEvent.value = success.blobCalendarEventList.first.calendarEventList.first;
+    _displayingEventBlobId = success.blobCalendarEventList.first.blobId;
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:1482]: Only need use `BlobCalendarEvent` replace to `CalendarEvent` & `DisplayingEventBlobId`

Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 9]
```diff
@@ -127,7 +127,10 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
   final emailContents = RxnString();
   final attachments = <Attachment>[].obs;
   final calendarEvent = Rxn<CalendarEvent>();
-  final eventActions = <EventAction>[].obs;
+  final eventActions = [
+    EventAction(EventActionType.yes, ''),
+    EventAction(EventActionType.no, ''),
+    EventAction(EventActionType.maybe, '')];
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:134]: Now we don't need it anymore

File: lib/features/email/domain/usecases/parse_calendar_event_interactor.dart [hunk 10]
```diff
@@ -22,26 +20,10 @@ class ParseCalendarEventInteractor {
     try {
       yield Right(ParseCalendarEventLoading());
 
-      final listResult = await Future.wait(
-        [
-          _calendarEventRepository.parse(accountId, blobIds),
-          _calendarEventRepository.getListEventAction(emailContents),
```
Comment (dab246) [inline lib/features/email/domain/usecases/parse_calendar_event_interactor.dart:28]: IMO, also should delete `getListEventAction` function in repository and datasource when not used

File: lib/features/email/presentation/controller/single_email_controller.dart (hunk 9 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

File: lib/features/email/domain/usecases/parse_calendar_event_interactor.dart (hunk 10 above)
Comment (tddang-linagora) [reply lib/features/email/domain/usecases/parse_calendar_event_interactor.dart:28]: Done

**Review by hoangdat — APPROVED** [review]

**Review by dab246 — APPROVED** [review]
Please rebase

**Pull request merged**